File: prototype/code.cpp
#include <iostream>
#include <string>
using namespace std;

// Prototype interface
class Prototype {
public:
    virtual Prototype* clone() = 0;
    virtual void describe() = 0;
    virtual ~Prototype() {}
};

// Employee class
class Employee : public Prototype {
    string name;
    string role;
    string department;
    int salary;

public:
    Employee(string name, string role, string department, int salary)
        : name(name), role(role), department(department), salary(salary) {}

    Prototype* clone() override {
        return new Employee(*this);
    }

    void setName(string newName) { name = newName; }
    void setSalary(int newSalary) { salary = newSalary; }

    void describe() override {
        cout << "Name: " << name << ", Role: " << role
             << ", Dept: " << department << ", Salary: " << salary << endl;
    }
};

// Client code
int main() {
    // Base template employee
    Employee* baseDev = new Employee("Template", "Software Engineer", "IT", 60000);

    // Hiring new employees by cloning
    Employee* emp1 = dynamic_cast<Employee*>(baseDev->clone());
    emp1->setName("Abhinav");

    Employee* emp2 = dynamic_cast<Employee*>(baseDev->clone());
    emp2->setName("Rohan");
    emp2->setSalary(65000);

    baseDev->describe();
    emp1->describe();
    emp2->describe();

    delete baseDev;
    delete emp1;
    delete emp2;
    return 0;
}

File: prototype/code.ts
// Prototype interface
interface Prototype {
  clone(): Prototype;
}

// Employee class
class Employee implements Prototype {
  name: string;
  role: string;
  department: string;
  salary: number;

  constructor(name: string, role: string, department: string, salary: number) {
    this.name = name;
    this.role = role;
    this.department = department;
    this.salary = salary;
  }

  clone(): Employee {
    return new Employee(this.name, this.role, this.department, this.salary);
  }

  describe() {
    console.log(
      `Name: ${this.name}, Role: ${this.role}, Dept: ${this.department}, Salary: ${this.salary}`
    );
  }
}

// Example usage
const baseDev = new Employee("Template", "Software Engineer", "IT", 60000);

// Hiring new employees by cloning
const emp1 = baseDev.clone();
emp1.name = "Abhinav";

const emp2 = baseDev.clone();
emp2.name = "Rohan";
emp2.salary = 65000; // salary tweak

baseDev.describe(); // Name: Template, Role: Software Engineer, Dept: IT, Salary: 60000
emp1.describe();    // Name: Abhinav, Role: Software Engineer, Dept: IT, Salary: 60000
emp2.describe();    // Name: Rohan, Role: Software Engineer, Dept: IT, Salary: 65000
